src/v/cluster/cluster_utils.cc

```diff
@@ -20,35 +20,39 @@
 
 #include <seastar/core/future.hh>
 
+#include <algorithm>
 #include <chrono>
 
 namespace cluster {
 patch<broker_ptr> calculate_changed_brokers(
-  std::vector<broker_ptr> new_list, std::vector<broker_ptr> old_list) {
+  const std::vector<broker_ptr>& new_list,
+  const std::vector<broker_ptr>& old_list) {
     patch<broker_ptr> patch;
-    auto compare_by_id = [](const broker_ptr& lhs, const broker_ptr& rhs) {
-        return *lhs < *rhs;
-    };
-    // updated/added brokers
-    std::sort(new_list.begin(), new_list.end(), compare_by_id);
-    std::sort(old_list.begin(), old_list.end(), compare_by_id);
-    std::set_difference(
-      std::cbegin(new_list),
-      std::cend(new_list),
-      std::cbegin(old_list),
-      std::cend(old_list),
-      std::back_inserter(patch.additions),
-      [](const broker_ptr& lhs, const broker_ptr& rhs) {
-          return *lhs != *rhs;
-      });
-    // removed brokers
-    std::set_difference(
-      std::cbegin(old_list),
-      std::cend(old_list),
-      std::cbegin(new_list),
-      std::cend(new_list),
-      std::back_inserter(patch.deletions),
-      compare_by_id);
+
+    for (auto br : new_list) {
+        auto it = std::find_if(
+          old_list.begin(), old_list.end(), [&br](const broker_ptr& ptr) {
+              return br->id() == ptr->id()
+                     && br->kafka_advertised_listeners()
+                          == ptr->kafka_advertised_listeners()
+                     && br->rpc_address() == ptr->rpc_address();
+          });
+
+        if (it == old_list.end()) {
+            patch.additions.push_back(br);
+        }
+    }
+
+    for (auto br : old_list) {
+        auto it = std::find_if(
+          new_list.begin(), new_list.end(), [&br](const broker_ptr& ptr) {
+              return br->id() == ptr->id();
```

Comment:
I think a comment about why the conditional here isn't identical with the one above would be useful?